llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -3165,9 +3166,13 @@ Instruction *InstCombinerImpl::foldICmpAddConstant(ICmpInst &Cmp,
     // If there is overflow, the result must be true or false.
     // TODO: Can we assert there is no overflow because InstSimplify always
     // handles those cases?
-    if (!Overflow)
+    if (!Overflow) {
```

Comment:
This looks incorrect as implemented, because the isSigned() check above still uses the original predicate. It would be good to add a test for a case where the ssub_ov vs usub_ov distinction is relevant.

Reply:
The overall implementation approach here is iffy, because for the nuw+nsw+samesign case we could use both the signed and unsigned predicate, so we should try both, with a preference for unsigned.

Reply:
1. You are right, let me fix it.
2. I don't get the second comment, can an instruction have both nsw and nuw at the same time how would that work?

Reply:
Interesting just noticed the tests related to nuw and nsw failing, let me check on this.

Reply:
> I don't get the second comment, can an instruction have both nsw and nuw at the same time how would that work?

Yes, you can have both nsw and nuw on an instruction. It's pretty common. E.g. add i32 1, 1 is both nuw and nsw.